### PR TITLE
[diffusion]: optimize zimage with rmsnorm + add + (rmsnorm)

### PR DESCRIPTION
**Pull request by attack204**
# E2E IMPROMENT

10s/13% faster

```bash
python python/sglang/multimodal_gen/benchmarks/compare_perf.py baseline.json new.json
### Performance Comparison Report
```

#### 1. High-level Summary
| Metric | Baseline | New | Diff | Status |
| :--- | :--- | :--- | :--- | :--- |
| **E2E Latency** | 71805.84 ms | 62156.90 ms | **-9648.94 ms (-13.4%)** | ✅ |
| **Throughput** | 0.01 req/s | 0.02 req/s | - | - |

#### 2. Stage Breakdown
| Stage Name | Baseline (ms) | New (ms) | Diff (ms) | Diff (%) | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| InputValidationStage | 0.06 | 0.06 | -0.01 | -9.4% | ⚪️ |
| TextEncodingStage | 929.59 | 686.98 | -242.61 | -26.1% | 🟢 |
| ConditioningStage | 0.03 | 0.03 | -0.00 | -7.0% | ⚪️ |
| TimestepPreparationStage | 2.25 | 2.12 | -0.13 | -5.7% | ⚪️ |
| LatentPreparationStage | 9.30 | 4.89 | -4.41 | -47.4% | ⚪️ |
| DenoisingStage | 69765.61 | 60430.47 | -9335.14 | -13.4% | 🟢 |
| DecodingStage | 1061.44 | 1000.04 | -61.40 | -5.8% | ⚪️ |

## DETAIL PERF

TODO

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/diffusion/fused_norm_scale_shift.cuh [hunk 1]
```diff
@@ -0,0 +1,1816 @@
...
+#include <sgl_kernel/tensor.h>   // For TensorMatcher, SymbolicSize, SymbolicDevice
+#include <sgl_kernel/utils.cuh>  // For LaunchKernel
+#include <sgl_kernel/utils.h>    // For div_ceil, RuntimeCheck
+
+#include <cuda_fp16.h>
+#include <dlpack/dlpack.h>
+#include <tvm/ffi/container/tensor.h>
+#include <tvm/ffi/optional.h>
+
+#include "cutlass/numeric_types.h"
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/diffusion/fused_norm_scale_shift.cuh:11]: Can we try to eliminate this dependency? It seems only numeric types are used.

File: python/sglang/multimodal_gen/runtime/layers/layernorm.py [hunk 2]
```diff
@@ -469,3 +469,140 @@ def apply_qk_norm(
     q_out = q_norm(q.view(-1, head_dim)).view(q_shape)
     k_out = k_norm(k.view(-1, head_dim)).view(k_shape)
     return q_out, k_out
+
+
+# ========== New Fused RMSNorm Classes for ZImage ==========
+
+class FusedRMSNormAddRMSNorm(nn.Module):
```
Comment (yingluosanqian) [inline python/sglang/multimodal_gen/runtime/layers/layernorm.py]: Could we use `FusedRMSNormAddRMSNorm(CustomOp)` instead? This way, we wouldn’t need to manually check `is_cuda`, since `CustomOp` already implements the dispatch logic, as shown below:
``` python
    def dispatch_forward(self) -> Callable:
        if current_platform.is_cuda():
            return self.forward_cuda
        elif current_platform.is_hip():
            return self.forward_hip
        elif current_platform.is_npu():
            return self.forward_npu
        elif current_platform.is_xpu():
            return self.forward_xpu
        else:
            return self.forward_native
```

**Comment by yingluosanqian** [comment]
After this [MR](https://github.com/sgl-project/sglang/pull/16452)
 is merged, we could pass `dtype` as a C++ template parameter, eliminating the need for runtime dtype dispatch.

``` cuda
if (dtype.code == kDLFloat && dtype.bits == 32) {
    dispatch_ipt(DTypeTag<float4, float>{});
  } else if (dtype.code == kDLFloat && dtype.bits == 16) {
    dispatch_ipt(DTypeTag<half4, half>{});
  } else if (dtype.code == kDLBfloat && dtype.bits == 16) {
    dispatch_ipt(DTypeTag<bf16_4, cutlass::bfloat16_t>{});
  }
```

**Comment by DarkSharpness** [comment]
What's the relationship of this PR and #14717?

It seems there's many redundant code between these 2 PRs. We may need to maximize code reuse and avoid too many similar code.

**Comment by yingluosanqian** [comment]
> What's the relationship of this PR and #14717?
> 
> It seems there's many redundant code between these 2 PRs. We may need to maximize code reuse and avoid too many similar code.

We have a [kernel fusion plan](https://acntwz63wyi6.feishu.cn/wiki/XRfdwY9jFii2MKkTMGwcRJBhnuf?from=from_copylink) for diffusion kernel. Many of which involve `norm` operation. 

I summarized how we currently end up generating a large number of kernel source variants:
1. Some kernels must handle both 3D and 4D input shapes.
2. When N is smaller than a certain threshold, placing the input in shared memory yields better performance.
3. Different kernels have different fusion patterns. For example, some kernels fuse `norm + scale + shift`, while others fuse `norm + tanh + scale + shift`, or `norm + add + norm`. **This is especially problematic, and we may need a better abstraction to maximize code reuse.**
